### PR TITLE
[WIP] Support source and target features

### PR DESCRIPTION
**Pull request by anderleich**
**ATTENTION! I'm opening a new PR as v3.0 branch has already been merged in `master`**
**The previous PR was https://github.com/OpenNMT/OpenNMT-py/pull/2227/. Closed**

This PR intends to add target features support to OpenNMT-py v3.0. All the code has been adapted for this new version.

Both source and target features support has been refactored for a more simplified handling of features. The way features are passed to the system has been changed and now features are appended to the actual textual data instead of providing a separate file. This also simplifies the way features are passed during inference and to the server. It uses the special character `￨` as a feature separator, as in the previous versions of the OpenNMT framework. For instance:
```
 I￨1￨3 love￨0￨1 eating￨0￨1 pizza￨0￨1
```
I've also added a way to provide default values for features. This can be really useful when mixing task specific data (with features) with general data which has not been annotated. Additionally, the `filterfeats` transform is no longer required and features are checked in the corpus loading process.

A YAML configuration file would look like this:
```
data:
    train:
        path_src: src_with_features.txt  #  I￨1￨3 love￨0￨1 eating￨0￨1 pizza￨0￨1
        path_tgt: tgt_with_features.txt  #  Me￨1 gusta￨0 comer￨0 pizza￨0
        transforms: [onmt_tokenize, inferfeats, filtertoolong]
    valid:
        path_src: src_with_features.txt
        path_tgt: tgt_with_features.txt
        transforms: [onmt_tokenize, inferfeats]

save_data: ./data
n_sample: -1

# # Vocab opts
src_vocab: data.vocab.src
tgt_vocab: data.vocab.tgt
n_src_feats: 2
n_tgt_feats: 1
src_feats_defaults: "0￨1"
tgt_feats_defaults: "1"
feat_merge: "sum"
```
For now, I've made the necessary changes in the code for vocabulary generation. That is, to make `onmt_build_vocab` work.

### CODE REVIEW

**Comment by vince62s** [comment]
@anderleich is there anything (at least source features) that we can test so far ? are you still working on this ?

**Comment by anderleich** [comment]
Hi @vince62s ,

Yes, I'm still working on this. I haven't had much time for this lately, but I'm on the right path. I've already managed to modify the vocabulary building and training scripts. These both work for source and target features. I need to ensure nothing is broken for other use cases (some debugging to make tests pass). After that the only thing left would be to modify the inference (onmt_translate and the server).

**Comment by anderleich** [comment]
Hi @vince62s ,

I've made code changes to make other configurations pass the tests. I still have some issues with the beam search translation and LM generation in the tests. Maybe you can shed some light on this? However, I think we are at a stage where we can begin to disccuss and test the changes I've made so far.

Note: I took the changes @francoishernandez made in this PR https://github.com/OpenNMT/OpenNMT-py/pull/1710 as a guide.

File: onmt/inputters/inputter.py [hunk 1]
```diff
@@ -167,9 +187,9 @@ def dict_to_vocabs(vocabs_dict):
...
-        vocabs['src_feats'] = {}
-        for feat in vocabs_dict['src_feats'].keys():
-            vocabs['src_feats'][feat] = \
+        vocabs['src_feats'] = []
+        for feat_vocab in vocabs_dict['src_feats']:
+            vocabs['src_feats'].append(
                 pyonmttok.build_vocab_from_tokens(
-                    vocabs_dict['src_feats'][feat])
+                    feat_vocab))
     return vocabs
```
Comment (vince62s) [inline onmt/inputters/inputter.py:195]: don't we need to add a section for vocabs['tgt_feats'] as well ?

File: onmt/inputters/text_utils.py [hunk 2]
```diff
@@ -16,15 +41,22 @@ def text_sort_key(ex):
...
     maybe_example['src'] = {"src": ' '.join(maybe_example['src'])}
     # Make features part of src like
-    # {'src': {'src': ..., 'feat1': ...., 'feat2': ....}}
-    if 'src_feats' in maybe_example:
-        for feat_name, feat_value in maybe_example['src_feats'].items():
-            maybe_example['src'][feat_name] = ' '.join(feat_value)
-        del maybe_example['src_feats']
+    # {'src': {'src': ..., 'feats': ....}
+    maybe_example['src']['feats'] = \
+        [' '.join(x) for x in maybe_example["src_feats"]]
```
Comment (vince62s) [inline onmt/inputters/text_utils.py:46]: are we 100% sure the key 'feats' always exists ?

File: onmt/inputters/text_utils.py [hunk 3]
```diff
@@ -16,15 +41,22 @@ def text_sort_key(ex):
...
     if maybe_example['tgt'] is not None:
         maybe_example['tgt'] = {'tgt': ' '.join(maybe_example['tgt'])}
+        maybe_example['tgt']['feats'] = \
+            [' '.join(x) for x in maybe_example["tgt_feats"]]
     if 'align' in maybe_example:
         maybe_example['align'] = ' '.join(maybe_example['align'])
+
+    # Delete these fields as they are no longer needed
+    del maybe_example['src_original']
+    del maybe_example['tgt_original']
```
Comment (vince62s) [inline onmt/inputters/text_utils.py:56]: I was actually wondering if we may need to keep those for future sanity checks, especially in the server.

File: onmt/inputters/text_utils.py [hunk 4]
```diff
@@ -81,12 +111,22 @@ def numericalize(vocabs, example):
...
-            src_feat = example['src'][featname].split()
-            vf = vocabs['src_feats'][featname]
-            # we'll need to change this if we introduce tgt feat
-            numeric['src'][featname] = vf(src_feat)
+        numeric_feats = []
+        for feat_vocab, feat in zip(
+                vocabs['src_feats'], example['src']['feats']):
+            numeric_feats.append(feat_vocab(feat.split()))
+        numeric['src']['feats'] = numeric_feats
+
```
Comment (vince62s) [inline onmt/inputters/text_utils.py:121]: ok here it gets tricky.
don't you think we need to use numeric['src']['feats_ids'] for clarity ?

File: onmt/inputters/inputter.py (hunk 1 above)
Comment (anderleich) [reply onmt/inputters/inputter.py:195]: Sure! Changed in the next commit

**Comment by vince62s** [comment]
ok, I have made a first pass. I need to clone your repo and test locally on my system. Will try to do it asap.
Jusr  make sure you have rebased and added the few things.

File: onmt/inputters/text_utils.py (hunk 2 above)
Comment (anderleich) [reply onmt/inputters/text_utils.py:46]: `maybe_example["src_feats"]` will be an empty list in those cases, and so will be `maybe_example['src']['feats']`. Maybe it's better to remove it from the dict?

File: onmt/inputters/text_utils.py (hunk 3 above)
Comment (anderleich) [reply onmt/inputters/text_utils.py:56]: What do you have in mind?

**Comment by anderleich** [comment]
I think that having dictionary like examples was the source of many incosistencies with source or target features. Therefore, I've added a new class `onmt.inputters.example.Example` to store all the necessary information for the input data, as well as to numericalize and transform the examples. This new class handles all the possible combinations for the input data, that is, the existence of the target sentence, source features, target features, alignments, the copy mechanism...
I could not come up with a reason for keeping the dict like examples, do you?

**Comment by vince62s** [comment]
@anderleich I don't mind discussing this new stuff but I really think it is too many changes for a single PR, especially when having done a first review.
Might be good to settle a bit with some changes and step by step include new concepts.

**Comment by anderleich** [comment]
I agree that adding the `Example` class resulted in many small changes in the code, specially in the tranforms. However, I think that overall it helped solving some of the inconsistencies with data I was dealing with, speacially now that I've also added target features.

**Comment by vince62s** [comment]
I get it but the issue is that we have plenty of missing unit tests a bit everywhere so we'll never be sure that it does not break things. I really would prefer to do things in at least two steps.

**Comment by anderleich** [comment]
Do you know a quick method to split the changes in two steps? Quicker than typing everything again...
I think is worth a try, if I get to pass all the unit tests implemented so far.

**Comment by vince62s** [comment]
You need to
1) create a new clean branch from master (don't forget to pull master too)
2) then you cherry pick commits from your working branch (google this you'll find,not difficult)

**Comment by anderleich** [comment]
Hi @vince62s ,

This is what I've finally planned on this. I will submit 3 different PRs to keep changes simpler and make reviews easier:

1. Restore back source features
2. Add the Example class to set the ground for the more complex scenario with target features
3. Add target features support

What do you think?

**Comment by vince62s** [comment]
We can try, let's do step 1. and we'll see how it goes.

**Comment by anderleich** [comment]
@vince62s I've created a new PR https://github.com/OpenNMT/OpenNMT-py/pull/2308 to restore back source features. All the test are passing. I'm planning to carry out some more checks to ensure everything is working as expected but overall the source features functionality is back. You can start reviewing the code.

TODO: I need to update the docs

**Pull request closed without merging**